New Docs Website / Accessibility Foundations page (#822)
hashicorp/design-system: website/docs/foundations/accessibility/index.md

---

> title: Accessibility Foundations

Reviewer: ```suggestion
title: Accessibility
```
_Foundations_ feels redundant given we're in the foundations section already

Author: I wanted to make it "Foundations: Accessibility" but I can't put colons in frontmatter. Any ideas? Goal is to have a unique page title. Unless we are merging page titles?

Reviewer: unique throughout the entire site?

Author: Yes; ideally we'll have something like `Page Name | Category Name | Site Name` (category/section name if it exists)

Reviewer 2: Does "title" here refer to what is displayed in the cover or the HTML `<title>` tag?

Reviewer 3: This is the title that appears in the page "cover". 
The page title (the one in the `<head>` comes from this function here:
https://github.com/hashicorp/design-system/blob/main/website/app/services/head-data.js#L9-L15
so we can customize it as we want (more or less, it depends on the data available at that level in the application).

Reviewer: 👍  i have a task in the backlog related to the titles not working correctly in certain cases. Can make this update at the same time.

2. It moves the focus to that message for the screen reader user, effectively resetting focus in Ember apps (similar to how a native web page/site works).
3. It provides a bypass mechanism so the user can skip to the page's primary content (see [https://www.w3.org/TR/WCAG20-TECHS/G1.html](https://www.w3.org/TR/WCAG20-TECHS/G1.html)). You can opt out of this if you want (see the `Options` section in the documentation for available options).

## Page Organization

All content should be contained in elements with [landmark roles](https://www.w3.org/TR/wai-aria/#landmark_roles). Typically this will be the `header`, `main`, and `footer` elements, but depending on your page layout, you might also be using `aside` and `nav` as direct descendants of the `body` element as well.

## Titles and Labels

### Page titles

Ensure that each page has a relevant page title. This is included in Ember applications by default, but older applications that have been updated over the years may need to manually add this in with [ember-page-title](https://github.com/ember-cli/ember-page-title).

### Headings

> Headings (h1-h6) should follow a logical order and not skip any levels. This can be tricky in dynamic applications so watch out for this one.

Author: ```suggestion
Headings (h1-h6) should follow a logical order and not skip any levels. Dynamic applications can make this challenging, so this may require extra attention in checking what renders to the browser.
```

> Headings (h1-h6) should follow a logical order and not skip any levels. This can be tricky in dynamic applications so watch out for this one.

Reviewer 4: How about something along these lines:
"This can be tricky in dynamic applications so watch out for this one."
=>
"Dynamic applications can make this challenging so may require extra attention."

### Accessible name

Not every element is allowed to use `aria-label`. Check the [element's role documentation](https://www.w3.org/TR/wai-aria/#role_definitions) to see what aria-attributes are permissible. When in doubt, ask in [#team-design-system](https://hashicorp.slack.com/archives/C7KTUHNUS)! As the saying goes, "no ARIA is better than bad ARIA!"

## Focus Management

There are some key points to remember about focus management:

1. Focus should return from whence it came. If a button toggled a modal window, closing that modal window should return focus to the button.
2. Trap the focus! Focus should remain in a modal or other popup when it is open. Try [ember-focus-trap](https://github.com/josemarluedke/ember-focus-trap).
3. Focus order should match DOM order.

## Automated Testing

Use available automated tooling for static and dynamic analysis. Ember apps can use [ember-template-lint](https://github.com/ember-template-lint/ember-template-lint) for static analysis and [ember-a11y-testing](https://github.com/ember-a11y/ember-a11y-testing) for dynamic analysis. Additionally, consider adding HTML validation to your CI.

> 

Reviewer 4: Just a note: If we are recommending others consider validating their HTML then we may want to look at making sure our components use valid markup as well. (For example currently we use div tags inside our button component which is invalid.)

Author: I agree! I think we have an issue in our backlog.

## Manual Testing

- *Keyboard support:* ensure that your application is keyboard navigable. Everything that you can do with a mouse should have an equivalent experience with a keyboard!

> - *Browser extensions:* there are a few browser extensions that can help you with accessibility. Check out Microsoft's [Accessibility Insights](https://accessibilityinsights.io/) and Deque's [aXe](https://www.deque.com/axe/browser-extensions/) are two popular extensions that can help you check for accessibility issues right in your browser!

Reviewer 2: ```suggestion
- *Browser extensions:* there are a few browser extensions that can help you with accessibility. Microsoft's [Accessibility Insights](https://accessibilityinsights.io/) and Deque's [aXe](https://www.deque.com/axe/browser-extensions/) are two popular extensions that can help you check for accessibility issues right in your browser.
```

> - *Safari and VoiceOver:* your development environment has a screen reader built in! Become familiar with VoiceOver use it to test your application with Safari, especially if you are creating a non-standard UI and want to make sure that it will be inclusive for our users.

Reviewer 2: 1. I wonder if these should be headings instead of emphasized text so they're more scannable? 
2. Sorry, I know this is lame to call out, but I'm not sure if these instances warrant the exclamation. 
3. Do we know if all of our consumers are on Mac? Do we have any folks using other operating systems? Can we add details for those too?
4. There are a few plugins we can add to this list for Figma, I'll send them over tomorrow.

Author: @heatherlarsen Ideally I'd like to make two tabs, one for designers and one for developers. WDYT?

Reviewer 2: Yea, that makes sense to me.
